akka-http-core/src/main/scala/akka/http/javadsl/model/ws/WebSocket.scala

```diff
@@ -19,4 +22,21 @@ object WebSocket {
       case Some(header) ⇒ header.handleMessagesWith(handler)
       case None         ⇒ HttpResponse.create().withStatus(StatusCodes.BAD_REQUEST).withEntity("Expected WebSocket request")
     }
+
+  /**
+   * If a given request is a WebSocket request a response accepting the request is returned using the given handler to
+   * handle the WebSocket message stream, as well as a future materialized value. If the request wasn't a WebSocket
+   * request a response with status code 400 is returned, as well as a failed future.
+   */
+  def handleWebSocketRequestWithMat[Mat](
+    request:     HttpRequest,
+    handler:     Graph[FlowShape[Message, Message], Mat],
+    subprotocol: String): Pair[HttpResponse, Future[Mat]] =
+    request.asScala.header[UpgradeToWebSocket] match {
+      case Some(header) ⇒ header.handleMessagesWithMat(handler, subprotocol)
```

Comment:
We'd like to prevent duplication of functional code, so either `handleWebSocketRequest` should use the new method or an abstraction should be created to handle both cases.

Reply:
@jrudolph i'm not sure i understand, are you referring to `handleWebSocketRequestWith`? and what is being duplicated?